examples_tests/01.HelloWorld/main.cpp

```diff
@@ -286,13 +302,236 @@ int main()
 	sc_params.queueFamilyIndices = queueFamilyIndices;
 	sc_params.imageSharingMode = imageSharingMode;
 	sc_params.preTransform = preTransform;
+	sc_params.imageUsage = asset::IImage::EUF_COLOR_ATTACHMENT_BIT;
 
 	core::smart_refctd_ptr<video::ISwapchain> swapchain = device->createSwapchain(std::move(sc_params));
 
-	while (true)
+	// Create render pass
+	video::IGPURenderpass::SCreationParams::SAttachmentDescription attachmentDescription;
+	attachmentDescription.format = surfaceFormat.format; // this should be same as the imageview used for this attachment
+	attachmentDescription.samples = asset::IImage::ESCF_1_BIT;
+	attachmentDescription.loadOp = video::IGPURenderpass::ELO_CLEAR; // when the first subpass begins with this attachment, clear its color and depth components
+	attachmentDescription.storeOp = video::IGPURenderpass::ESO_STORE; // when the last subpass ends with this attachment, store its results
+	attachmentDescription.initialLayout = asset::EIL_UNDEFINED;
+	attachmentDescription.finalLayout = asset::EIL_PRESENT_SRC_KHR;
+
+	video::IGPURenderpass::SCreationParams::SSubpassDescription subpassDescription;
+	subpassDescription.flags = video::IGPURenderpass::ESDF_NONE;
+	subpassDescription.pipelineBindPoint = asset::EPBP_GRAPHICS;
+	subpassDescription.inputAttachmentCount = 0u;
+	subpassDescription.inputAttachments = nullptr;
+
+	video::IGPURenderpass::SCreationParams::SSubpassDescription::SAttachmentRef colorAttRef;
+	{
+		colorAttRef.attachment = 0u;
+		colorAttRef.layout = asset::EIL_COLOR_ATTACHMENT_OPTIMAL;
+	}
+	subpassDescription.colorAttachmentCount = 1u;
+	subpassDescription.colorAttachments = &colorAttRef;
+	subpassDescription.resolveAttachments = nullptr;
+	subpassDescription.depthStencilAttachment = nullptr;
+	subpassDescription.preserveAttachmentCount = 0u;
+	subpassDescription.preserveAttachments = nullptr;
+
+	video::IGPURenderpass::SCreationParams renderPassParams;
+	renderPassParams.attachmentCount = 1u;
+	renderPassParams.attachments = &attachmentDescription;
+	renderPassParams.dependencies = nullptr;
+	renderPassParams.dependencyCount = 0u;

```

Comment:
@achalpandeyy check with Vulkan discord #synchronisation channel, from my reading of the spec it looks like there are implicit execution dependencies between commands outside a renderpass, and inside a subpass, but no ordering w.r.t. the subpasses themselves, so I wonder if you need a dependency between <VK_SUBPASS_EXTERNAL,0> and <0,VK_SUBPASS_EXTERNAL> ?

Reply:
ok I've done some reading

> If there is no subpass dependency from VK_SUBPASS_EXTERNAL to the first subpass that uses an attachment, then an implicit subpass dependency exists from VK_SUBPASS_EXTERNAL to the first subpass it is used in.

> The implicit subpass dependency only exists if there exists an automatic layout transition away from initialLayout.

So if in your renderpasses description you include a layout transition from PRESENT to COLOR_ATTACHMENT you get this dependency for free
```cpp
VkSubpassDependency implicitDependency = {
    .srcSubpass = VK_SUBPASS_EXTERNAL;
    .dstSubpass = firstSubpass; // First subpass attachment is used in
    .srcStageMask = VK_PIPELINE_STAGE_NONE_KHR;
    .dstStageMask = VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
    .srcAccessMask = 0;
    .dstAccessMask = VK_ACCESS_INPUT_ATTACHMENT_READ_BIT |
                     VK_ACCESS_COLOR_ATTACHMENT_READ_BIT |
                     VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT |
                     VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_READ_BIT |
                     VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
    .dependencyFlags = 0;
};
```

likewise if your `finalLayout` is PRESENT then you get this one for free
```cpp
VkSubpassDependency implicitDependency = {
    .srcSubpass = lastSubpass; // Last subpass attachment is used in
    .dstSubpass = VK_SUBPASS_EXTERNAL;
    .srcStageMask = VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
    .dstStageMask = VK_PIPELINE_STAGE_NONE_KHR;
    .srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT |
                     VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT;
    .dstAccessMask = 0;
    .dependencyFlags = 0;
};
```

so you get two halfs of 2 different split barriers.

You dont need the first half (availability before renderpass) because present doesnt write to the image memory.

The last half is implicit because a submit/present imply a device scope visibility op AFAIK.